File: src/http-client.ts
import type { IngestErrorPayload, HttpErrorReportingResult, ErrorReportingOptions } from "./types";

export class HttpErrorReportingClient {
  private convexUrl: string;
  private convexSecret: string;
  private options: Required<Pick<ErrorReportingOptions, 'maxRetries' | 'retryDelay' | 'timeout' | 'environment' | 'defaultContext'>>;
  
  constructor(convexUrl: string, convexSecret: string, options: ErrorReportingOptions = {}) {
    if (!convexUrl) {
      throw new Error("Convex URL is required for HTTP transport");
    }
    if (!convexSecret) {
      throw new Error("Convex API secret is required for HTTP transport");
    }

    this.convexUrl = convexUrl.replace(/\/$/, ''); // Remove trailing slash
    this.convexSecret = convexSecret;
    
    this.options = {
      maxRetries: options.maxRetries ?? 3,
      retryDelay: options.retryDelay ?? 1000,
      timeout: options.timeout ?? 10000,
      environment: options.environment ?? 'production',
      defaultContext: options.defaultContext ?? {}
    };
  }

  /**
   * Report an error with a known error code via HTTP
   */
  async reportKnownError(
    companyCode: string,
    errorCode: string,
    message: string,
    context?: Record<string, any>
  ): Promise<HttpErrorReportingResult> {
    const payload = {
      companyCode,
      errorCode,
      message,
      context: { ...this.options.defaultContext, ...context },
      environment: this.options.environment
    };

    return this.makeRequest('/report-known-error', payload);
  }

  /**
   * Report a service error for automatic classification via HTTP
   */
  async reportServiceError(
    companyCode: string,
    service: string,
    message: string,
    context?: Record<string, any>
  ): Promise<HttpErrorReportingResult> {
    const payload = {
      companyCode,
      service,
      message,
      context: { ...this.options.defaultContext, ...context },
      environment: this.options.environment
    };

    return this.makeRequest('/report-service-error', payload);
  }

  /**
   * Report an error with full context via HTTP
   */
  async reportError(payload: IngestErrorPayload): Promise<HttpErrorReportingResult> {
    const enrichedPayload = {
      ...payload,
      context: { ...this.options.defaultContext, ...payload.context },
      environment: payload.environment ?? this.options.environment
    };

    return this.makeRequest('/report-error', enrichedPayload);
  }

  private async makeRequest(endpoint: string, payload: any): Promise<HttpErrorReportingResult> {
    let lastError: Error | null = null;
    
    for (let attempt = 1; attempt <= this.options.maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.options.timeout);

        const response = await fetch(`${this.convexUrl}${endpoint}`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${this.convexSecret}`
          },
          body: JSON.stringify(payload),
          signal: controller.signal
        });

        clearTimeout(timeoutId);

        const result = await response.json() as HttpErrorReportingResult;

        if (!response.ok) {
          throw new Error(`HTTP ${response.status}: ${result.message || 'Unknown error'}`);
        }

        return result;
      } catch (error) {
        lastError = error as Error;
        
        if (attempt < this.options.maxRetries) {
          const delay = this.options.retryDelay * Math.pow(2, attempt - 1);
          await new Promise(resolve => setTimeout(resolve, delay));
        }
      }
    }
    
    return {
      success: false,
      error: 'NETWORK_ERROR',
      message: lastError?.message ?? 'Failed to report error after retries'
    };
  }
}

File: python/railblocks_error_sdk/http_client.py
"""
HTTP client for the Railblocks Error SDK.

Python implementation that mirrors the TypeScript HttpErrorReportingClient exactly.
"""

import asyncio
import json
from typing import Dict, Any, Optional
import aiohttp
from .types import HttpErrorReportingResult, ErrorReportingOptions, IngestErrorPayload


class HttpErrorReportingClient:
    """Python wrapper for the TypeScript HttpErrorReportingClient."""
    
    def __init__(self, convex_url: str, convex_secret: str, options: Optional[ErrorReportingOptions] = None):
        if not convex_url:
            raise ValueError("Convex URL is required for HTTP transport")
        if not convex_secret:
            raise ValueError("Convex API secret is required for HTTP transport")
        
        self.convex_url = convex_url.rstrip('/')  # Remove trailing slash
        self.convex_secret = convex_secret
        
        # Set defaults matching TypeScript SDK
        options = options or ErrorReportingOptions()
        self.max_retries = options.max_retries or 3
        self.retry_delay = options.retry_delay or 1000
        self.timeout = options.timeout or 10000
        self.environment = options.environment or 'production'
        self.default_context = options.default_context or {}
    
    async def report_known_error(
        self,
        company_code: str,
        error_code: str,
        message: str,
        context: Optional[Dict[str, Any]] = None
    ) -> HttpErrorReportingResult:
        """Report an error with a known error code via HTTP."""
        payload = {
            "companyCode": company_code,
            "errorCode": error_code,
            "message": message,
            "context": {**self.default_context, **(context or {})},
            "environment": self.environment
        }
        
        return await self._make_request('/report-known-error', payload)
    
    async def report_service_error(
        self,
        company_code: str,
        service: str,
        message: str,
        context: Optional[Dict[str, Any]] = None
    ) -> HttpErrorReportingResult:
        """Report a service error for automatic classification via HTTP."""
        payload = {
            "companyCode": company_code,
            "service": service,
            "message": message,
            "context": {**self.default_context, **(context or {})},
            "environment": self.environment
        }
        
        return await self._make_request('/report-service-error', payload)
    
    async def report_error(self, payload: IngestErrorPayload) -> HttpErrorReportingResult:
        """Report an error with full context via HTTP."""
        enriched_payload = {
            "companyCode": payload.company_code,
            "message": payload.message,
            "errorCode": payload.error_code,
            "service": payload.service,
            "context": {**self.default_context, **(payload.context or {})},
            "environment": payload.environment or self.environment,
            "stackTrace": payload.stack_trace,
            "url": payload.url,
            "userAgent": payload.user_agent,
            "userId": payload.user_id,
            "sessionId": payload.session_id,
            "severity": payload.severity,
            "tags": payload.tags,
            "location": payload.location
        }
        
        return await self._make_request('/report-error', enriched_payload)
    
    async def _make_request(self, endpoint: str, payload: Dict[str, Any]) -> HttpErrorReportingResult:
        """Make HTTP request with retry logic matching TypeScript SDK exactly."""
        last_error: Optional[Exception] = None
        full_url = f"{self.convex_url}{endpoint}"
        
        for attempt in range(1, self.max_retries + 1):
            try:
                timeout = aiohttp.ClientTimeout(total=self.timeout / 1000)  # Convert to seconds
                
                async with aiohttp.ClientSession(timeout=timeout) as session:
                    async with session.post(
                        full_url,
                        headers={
                            'Content-Type': 'application/json',
                            'Authorization': f'Bearer {self.convex_secret}'
                        },
                        json=payload
                    ) as response:
                        # Get the raw response text first to handle JSON parsing properly
                        response_text = await response.text()
                        
                        try:
                            result_data = json.loads(response_text)
                        except Exception as json_error:
                            raise Exception(f"Failed to parse JSON response: {json_error}")
                        
                        if not response.ok:
                            error_msg = f"HTTP {response.status}: {result_data.get('message', 'Unknown error')}"
                            raise Exception(error_msg)
                        
                        return HttpErrorReportingResult(
                            success=result_data.get('success', False),
                            trigger_id=result_data.get('triggerId'),
                            message=result_data.get('message'),
                            error=result_data.get('error')
                        )
                        
            except Exception as error:
                last_error = error
                
                if attempt < self.max_retries:
                    delay = self.retry_delay * (2 ** (attempt - 1)) / 1000  # Convert to seconds
                    await asyncio.sleep(delay)
        
        return HttpErrorReportingResult(
            success=False,
            error='NETWORK_ERROR',
            message=str(last_error) if last_error else 'Failed to report error after retries'
        )


def create_http_error_client(convex_url: str, convex_secret: str, options: Optional[ErrorReportingOptions] = None) -> HttpErrorReportingClient:
    """Create a new HTTP error reporting client instance - matches TypeScript API exactly."""
    return HttpErrorReportingClient(convex_url, convex_secret, options) 